quartz/build.ts

```diff
@@ -110,19 +126,148 @@ async function startServing(
+      .on("unlink", (fp) => rebuildFromEntrypoint(fp, "delete", clientRefresh, buildData))
+  }
 
   return async () => {
     await watcher.close()
   }
 }
 
+async function partialRebuild(
+  filepath: string,
+  action: FileEvent,
+  clientRefresh: () => void,
+  buildData: BuildData, // note: this function mutates buildData
+) {
+  const { ctx, ignored, depGraphs, contentMap } = buildData
+  const { argv, cfg } = ctx
+  const toRemove = new Set<FilePath>()
+
+  // don't do anything for gitignored files
+  if (ignored(filepath)) {
+    return
+  }
+
+  const perf = new PerfTimer()
+  console.log(chalk.yellow("Detected change, rebuilding..."))
+
+  // UPDATE DEP GRAPH
+  const fp = joinSegments(argv.directory, toPosixPath(filepath)) as FilePath
+
+  const staticResources = getStaticResourcesFromPlugins(ctx)
+  let processedFile: ProcessedContent
+
+  switch (action) {
+    case "add":
+      // add to cache when new file is added
+      processedFile = (await parseMarkdown(ctx, [fp]))[0]
+      contentMap.set(fp, processedFile)
+
+      // update the dep graph by asking all emitters whether they depend on this file
+      for (const emitter of cfg.plugins.emitters) {
+        const emitterGraph = await emitter.getDependencyGraph(ctx, [processedFile], staticResources)
+        depGraphs[emitter.name].mergeEdgesForNode(emitterGraph, fp)
+      }
+      break
+    case "change":
+      // invalidate cache when file is changed
+      processedFile = (await parseMarkdown(ctx, [fp]))[0]
+      contentMap.set(fp, processedFile)
+
+      // only content files can have added/removed dependencies because of transclusions
+      if (path.extname(fp) === ".md") {
+        for (const emitter of cfg.plugins.emitters) {
+          // get new dependencies from all emitters for this file
+          const emitterGraph = await emitter.getDependencyGraph(
+            ctx,
+            [processedFile],
+            staticResources,
+          )
+          // merge the new dependencies into the dep graph
+          depGraphs[emitter.name].mergeEdgesForNode(emitterGraph, fp)
+        }
+      }
+      break
+    case "delete":
+      toRemove.add(fp)
+      break
+  }
+
+  if (argv.verbose) {
+    console.log(`Updated dependency graphs in ${perf.timeSince()}`)
+  }
+
+  // EMIT
+  perf.addEvent("rebuild")
+  let emittedFiles = 0
+  const destinationsToDelete = new Set<FilePath>()
+
+  for (const emitter of cfg.plugins.emitters) {
+    const depGraph = depGraphs[emitter.name]
+
+    // only call the emitter if it uses this file
+    if (depGraph.hasNode(fp)) {
+      // re-emit using all files that are needed for the downstream of this file
+      // eg. for ContentIndex, the dep graph could be:
+      // a.md --> contentIndex.json
+      // b.md ------^
+      //
+      // if a.md changes, we need to re-emit contentIndex.json,
+      // and supply [a.md, b.md] to the emitter
+      const upstreams = [...depGraph.getUpstreamsOfDownstreamLeafNodes(fp)] as FilePath[]
```

Comment:
`upstreams` needs to be _all_ the files (consider a folder listing, it would be wrong to only update the folder listing with changed files)

Reply:
though maybe we need to update `emitter.emit` to take in the dependency graph to figure out for itself what it needs to update

Reply:
The dependency graph the emitter returns should provide enough dependency info that we don't need to supply all the files. PR description has an example for ContentIndex where multiple files are required to create contentIndex.json, and `getUpstreamsOfDownstreamLeafNodes` gives us all the files needed. 

This approach also helps us do reload when embedded media (eg script) changes -- currently you have to restart quartz build to get the changed script. Though I'm not sure how to pass in assets like images/scripts into the emitters right now...